modules/packages/Sort.chpl

```diff
@@ -2606,104 +2818,190 @@ module TwoArrayPartitioning {
-              theLocaleId = tid;
+
+        // aka barrier
+
+        // now the data is all in A
+        /*if debugDist {
+          writef("after distribute, A is %xt\n", A[task.start..taskEnd]);
+        }*/
+
+        // Step 4: Add sub-tasks depending on if the bin is local or distributed
+        // still.
+        ref bktOwnerState = state.perLocale[bktLocId];
+
+        for bin in bktOwnerState.compat.bucketizer.getBinsToRecursivelySort() {
+          const binStart = if bin*nLocalesTotal > 0
+                            then bktOwnerState.globalEnds[bin*nLocalesTotal-1]
+                            else bktTask.start;
+          const binEnd = bktOwnerState.globalEnds[bin*nLocalesTotal+nLocalesTotal-1] - 1;
+          const binSize = binEnd - binStart + 1;
+          const binStartBit = bktOwnerState.compat.bucketizer.getNextStartBit(bktTask.startbit);
+          if binSize > 1 {
+            var small = false;
+            var theLocaleId = -1;
+
+            // Compute the regions on the same locale as the first, last
+            // elements in the bin.
+            const firstLoc = A.domain.dist.idxToLocale(binStart);
+            const lastLoc = A.domain.dist.idxToLocale(binEnd);
+            const onFirstLoc = A.localSubdomain(firstLoc)[binStart..binEnd];
+            const onLastLoc = A.localSubdomain(lastLoc)[binStart..binEnd];
+            var theLocale = firstLoc;
+            if onFirstLoc.size == binSize {
+              // case 1: all elements are on firstLoc
+              small = true;
+            } else if binSize <= state.distributedBaseCaseSize {
+              // case 2: size is small enough to do on 1 locale
+              small = true;
             }
-          }
+            theLocaleId = theLocale.id;
+            assert(A.targetLocales()[theLocaleId] == theLocale);
+
+            if debugDist then
+              writeln(bktLocId,
+                      " Recursive bin ", bin,
+                      " start = ", binStart,
+                      " size = ", binSize,
+                      " startbit = ", binStartBit,
+                      " small = ", small);
+
+            if small {
+              var small = new TwoArraySortTask(binStart, binSize, binStartBit,
+                                               true, true);
+
+              if debugDist then
+                writeln(bktLocId, " Adding small task ", small);
+
+              smallTasksPerLocale[theLocaleId].append(small);
+            } else {
+              // Create one or more distributed sorting tasks
+
+              // Distributed sorting tasks need to use the per-locale data
+              // structures allocated above...
+              // so we can have each locale involved in only one distributed
+              // sort task at a time.
+              // (we could consider relaxing this in various ways;
+              //  creating 2 state records being one option).
+
+              // Which locales will own this bucket?
```

Comment:
Is this a question to be answered still? Or is this saying "the following code computes who owns the bucket"?

Reply:
the comment is out of date - I've removed it (we do have 2 state records now)